File: exercises/bank-account/.meta/src/reference/java/BankAccount.java
class BankAccount {
    private int balance = 0;
    private boolean isClosed = true;

    void open() {
        isClosed = false;
    }

    void close() {
        isClosed = true;
    }

    synchronized int getBalance() throws BankAccountActionInvalidException {
        checkIfClosed();
        return balance;
    }

    synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        checkIfClosed();
        checkIfValidAmount(amount);

        balance += amount;
    }

    synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        checkIfClosed();
        checkIfValidAmount(amount);
        checkIfEnoughMoneyInAccount(amount);

        balance -= amount;
    }

    private void checkIfValidAmount(int amount) throws BankAccountActionInvalidException {
        if (amount < 0) {
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
        }
    }

    private void checkIfEnoughMoneyInAccount(int amount) throws BankAccountActionInvalidException {
        if (balance == 0) {
            throw new BankAccountActionInvalidException("Cannot withdraw money from an empty account");
        }
        if (balance - amount < 0) {
            throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");
        }
    }

    private void checkIfClosed() throws BankAccountActionInvalidException {
        if (isClosed) {
            throw new BankAccountActionInvalidException("Account closed");
        }
    }
}

File: go/bank-account/bank_account.go
package account

import "sync"

type Account struct {
	balance int64
	open    bool
	m       sync.RWMutex
}

func Open(initialDeposit int64) *Account {
	if initialDeposit < 0 {
		return nil
	}
	return &Account{initialDeposit, true, sync.RWMutex{}}
}

func (account *Account) Close() (int64, bool) {
	account.m.Lock()
	defer account.m.Unlock()
	if !account.open {
		return 0, false
	}
	account.open = false
	b := account.balance
	account.balance = 0
	return b, true
}

func (account *Account) Balance() (int64, bool) {
	account.m.RLock()
	defer account.m.RUnlock()
	return account.balance, account.open
}

func (account *Account) Deposit(amount int64) (int64, bool) {
	account.m.Lock()
	defer account.m.Unlock()
	if account.balance + amount < 0 {
		return account.balance, false
	}
	account.balance += amount
	return account.balance, account.open
}